ktwillison/app-code-test
# Feature/delete and add locations

## Description
This pull request implements the following:
* Adding a testable `WeatherLocationService` to facilitate request handling towards backend services
* To that end, adding `TargetType`s and mockable `URLSession`s and `URLSessionDataTask`s for use in defining and handling request and response logic for client-side services
* Test coverage for the same (for models, targets and the client handling)
* Add support for adding & deleting `WeatherLocations` in `WeatherLocationService` 
* Add UI for adding and deleting `WeatherLocations`
* Add a coordinator to manage navigation and view presentation logic


## Code test details
This was super fun, so thanks for the exercise! Was nice to play around in a bit of a different structure / architecture to what I'm working with most days. There are still a few things I'd like to clean up (you'll see TODOs in the code 😇), but I went on the upper end of time here (clocking in at about 5-6 hours) so I'll leave them for now -- hope that's not too painful / just gives us more to discuss! I've left my commits a bit more raw / unsquashed then I usually would so that you can get a sense of the code's progression as well, in case that's helpful.

<img src="https://user-images.githubusercontent.com/11081689/115550251-ecbe4f80-a2a9-11eb-9e66-3b29009978cd.png" width="200">

### Test plan
* Unit test coverage on business logic / `WeatherLocationService` as above
* Manual testing for view / view controller logic (though would love to automagic this), namely:
    * Swipe to delete a location; ensure that location is deleted in the UI, and when re-starting the app
    * Add a location with correct information; ensure the view is dismissed, and that it is added in the UI when returning to the main view
    * Add a location with missing information; ensure a validation error occurs and the view is not dismissed
    * Run the app on a small device, ensure scroll view on AddLocationPage behaves as expected.
* Hacky test coverage for error handling in the Controller / UI: 
    * Mock errors from all services (updating code directly), ensure errors are reported to the user when 
         1) refreshing the main view
         2) adding a new location 
         3) deleting a location
